src/main/java/org/voltdb/exportclient/KinesisFirehoseExportClient.java

```diff
@@ -0,0 +1,260 @@
+                    + " table " + source.tableName
+                    + " generation " + source.m_generation, CoreUtils.MEDIUM_STACK_SIZE);
+            m_decoder = builder.build();
+        }
+
+        private void validateStream() throws RestartBlockException, InterruptedException {
+            DescribeDeliveryStreamRequest describeHoseRequest = new DescribeDeliveryStreamRequest().
+                    withDeliveryStreamName(m_streamName);
+            DescribeDeliveryStreamResult  describeHoseResult = null;
+            String status = "UNDEFINED";
+            describeHoseResult = m_firehoseClient.describeDeliveryStream(describeHoseRequest);
+            status = describeHoseResult.getDeliveryStreamDescription().getDeliveryStreamStatus();
+            if("ACTIVE".equalsIgnoreCase(status)){
+                return;
+            }
+            else if("CREATING".equalsIgnoreCase(status)){
+                Thread.sleep(5000);
+                validateStream();
+            }
+            else {
+                LOG.error("Cannot use stream %s, responded with %s", m_streamName, status);
+                throw new RestartBlockException(true);
+            }
+        }
+
+        final void checkOnFirstRow() throws RestartBlockException {
+            if (!m_primed) try {
+                m_firehoseClient = new AmazonKinesisFirehoseClient(
+                        new BasicAWSCredentials(m_accessKey, m_secretKey));
+                m_firehoseClient.setRegion(m_region);
+                validateStream();
+            } catch (AmazonServiceException | InterruptedException e) {
+                LOG.error("Unable to instantiate a Amazon Kinesis Firehose client", e);
+                throw new RestartBlockException("Unable to instantiate a Amazon Kinesis Firehose client", e, true);
+            }
+            m_primed = true;
+        }
+
+        @Override
+        public boolean processRow(int rowSize, byte[] rowData) throws RestartBlockException {
+            if (!m_primed) checkOnFirstRow();
+            Record record = new Record();
+            try {
+                final ExportRowData rd = decodeRow(rowData);
+                String decoded = m_decoder.decode(null, rd.values);
+                record.withData(ByteBuffer.wrap(decoded.getBytes(StandardCharsets.UTF_8)));
+            } catch(IOException e) {
+                LOG.error("Failed to build record", e);
+                throw new RestartBlockException("Failed to build record", e, true);
+            }
+            // PutRecordBatchRequest can not contain more than 500 records
+            // And up to a limit of 4 MB for the entire request
+            if (((m_currentBatchSize + rowSize) > BATCH_SIZE_LIMIT) || (currentBatch.size() >= BATCH_NUMBER_LIMIT)) {
+                // roll to next batch
+                m_records.add(currentBatch);
+                m_currentBatchSize = 0;
+                currentBatch = new LinkedList<Record>();
```

Comment:
why not add it to m_records here? this way on onBlockComplete you dont need special handing jus go through m_records.